Combobox.js

```diff
@@ -641,65 +665,145 @@ define([
+		filter: function (inputText) {
+			// Escape special chars in search string, see
+		    // http://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex.
+		    var filterTxt = inputText.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
+		    if (this.filterMode === "startsWith") {
 				filterTxt = "^" + filterTxt;
 			} else if (this.filterMode === "is") {
 				filterTxt = "^" + filterTxt + "$";
 			} // nothing to add for "contains"
 
 			var rexExp = new RegExp(filterTxt, this.ignoreCase ? "i" : "");
-			this.list.query = (new Filter()).match(this.list.labelAttr, rexExp);
+			this.list.query = this.getQuery({
+				rexExp: rexExp,
+				inputText: inputText
+			});
+			if (this.source) {
+				this.list.source = this.source;
+			}
+
+			// Open popup in order to show the list's content (loading panel, no-items element or items).
+			this.openDropDown(true);
+		},
+
+		/**
+		 * Sets the new list's query.
+		 * This method can be overridden when using other store types.
+		 * The default implementation uses `dstore/Filter.match()`.
+		 * The matching is performed against the `list.labelAttr` or `list.labelFunc` attributes of
+		 * the data store items.
+		 * The method can be overridden for implementing other filtering strategies.
+		 * @protected
+		 * @returns {Object} New query to set to the list.
+		 */
+		getQuery: function (args) {
+			return (new Filter()).match(this.list.labelAttr || this.list.labelFunc, args.rexExp);
+		},
+
+		/**
+		 * Sets the initial list's query,
+		 * It can be one of a `function`, an `Object` or `null`.
+		 * If it is a function, then it's invoked and the list's query will get the return value assigned.
+		 * If it is an Object, it's assigned to the list's query directly.
+		 * If it is null, the list's query won't be set to its initial value.
+		 * It can be overridden depending of store used and the strategy to apply.
+		 */
+		getInitialQuery: {},
+
+		_setSelectedItems: function () {
+			if (this.list.source && this.list.renderItems && this.value !== "") {
+				var selectedItems = [],
+					presetItems = this.value instanceof Array && this.value.length >= 1 ? this.value : [this.value];
+				selectedItems = this.list.renderItems.filter(function (renderItem) {
+				    return presetItems.indexOf(this._getItemValue(renderItem)) >= 0;
+				}.bind(this));
+
+				this.list.selectedItems = selectedItems;
+				this._validateInput(false);
+			}
 		},
 
 		openDropDown: dcl.superCall(function (sup) {
-			return function () {
-				// Store the current selection, to be able to restore when pressing the
-				// cancel button. Used by ComboPopup. (Could spare it in situations when
-				// there is no cancel button, but not really worth.)
-				this._selectedItems = this.list.selectedItems;
+			return function (filteringInProgress) {
+				// Assign widget's source to the list's source.
+				if (this.source) {
+					this.list.source = this.source;
+				}
```

Comment:
I think we talked about always setting `this.list.query` and `this.list.source` at the same time (or in other words, defer setting `this.list.source` until you set `this.list.query`.   Maybe that's what's happening here, but it's hard to tell because I just see the line for `this.list.source`.

It would be clearer if you always put those two together.

Maybe you want a helper function to do that?  It could also handle `getInitialQuery`, specifically the part about branching depending on if `getInitialQuery` is an Object or a function.